scipy/sparse/_coo.py

```diff
@@ -652,7 +653,21 @@ def _matmul_vector(self, other):
 
 
     def _matmul_dispatch(self, other):
-        if self.ndim < 3:
+        if not (issparse(other) or isdense(other) or isscalarlike(other)):
+            # If it's a list or whatever, treat it like an array
+            other_a = np.asanyarray(other)
+
+            if other_a.ndim == 0 and other_a.dtype == np.object_:
+                # Not interpretable as an array; return NotImplemented so that
+                # other's __rmatmul__ can kick in if that's implemented.
+                return NotImplemented
+
+            try:
+                other.shape
+            except AttributeError:
+                other = other_a
+
+        if self.ndim < 3 and (np.isscalar(other) or other.ndim < 3):
```

Comment:
With scalars handled, this becomes only checking ndim<3 for self and other.

Reply:
Done at f01ba88873d4c333319e28c21d9f6f6e7cb62d16.